Add complete Megrez-MoE support: GGUF conversion + inference.

tamarPal opened the pull request:
>--Replaces closed PR #17052 (force-push issue). Only Megrez-MoE changes, fully tested.

## Architecture
- 64 routed experts (top-6) + 4 shared experts
- Sigmoid + bias gating
- 30 MoE layers, 2048-dim, 163K context

## Changes
1. **Conversion** (`convert_hf_to_gguf.py`): MegrezMoEModel class
   - Shared expert FFN: `hidden_size × 2.75 = 5632`
   - Gate bias mapping: `e_score_correction_bias` → `exp_probs_b`
   - Expert tensor merging (64 experts/layer)

2. **Architecture** (`constants.py`): MODEL_TENSORS.MEGREZ_MOE

3. **Inference** (`llama.cpp`): MoE FFN + graph memory fix

## Testing
-Real model: Infinigence/Megrez2-3x7B-A3B (13.92 GB → 14 GB GGUF)  
-Conversion: 372 tensors, all parameters correct  
-Inference: Coherent output, 5.77 tok/s  
-All 40 tests pass  

## Example
```bash
python3 convert_hf_to_gguf.py models/Megrez2-3x7B-A3B/
./build/bin/llama-cli -m megrez2-3x7b-a3b-f16.gguf -p "Hello" -n 50

Commit:
feat: Add Megrez-MoE architecture support

Implements complete support for Megrez-MoE (Mixture of Experts) models:

- Add LLM_ARCH_MEGREZ_MOE architecture enum and mappings
- Implement build_mergez_moe_ffn() with sigmoid+bias gating
- Add llm_build_megrez_moe class for full model graph construction
- Support 31-layer architecture (layer 0: dense FFN, layers 1-30: MoE)
- Implement expert sharing pattern with 64 experts, 6 used per token, 4 shared
- Load all model hyperparameters and 372 tensors correctly
- Configure NEOX RoPE type for proper positional encoding

Tested with Megrez2-3x7B-A3B_Q4_K_M.gguf model.
All 39 llama.cpp tests pass successfully.
Output verified to match infinigence/llama.cpp reference implementation.

Note: Use --no-warmup flag to avoid warmup memory allocation issue.

Commit:
fix: increase graph nodes for Megrez-MoE warmup

Megrez-MoE creates many intermediate tensors during MoE FFN construction:
- sigmoid, add, reshape (3x), get_rows, sum_rows, div, view_2d, mul_mat operations
- ggml_top_k internally calls ggml_argsort + ggml_view_4d (2 more tensors per layer)
- Each of 30 MoE layers creates ~35 intermediate tensors during graph construction

During warmup, the graph is built 3 times with different batch sizes, requiring
sufficient memory pool space for all intermediate tensors.

Add 4096 node overhead for LLM_ARCH_MEGREZ_MOE to accommodate these intermediate
tensors (30 layers × 35 tensors/layer ≈ 1050 nodes, doubled for safety margin).

This fixes the 'not enough space in the context's memory pool' error during warmup,
allowing Megrez-MoE to work without the --no-warmup flag.

Tested:
- All 39 tests pass
- Megrez-MoE works with warmup enabled (no crashes)
- Other models (e.g., Gemma-2) are unaffected
- Verified with outputs up to 100 tokens

Commit:
feat: adapt Megrez-MoE to new models/*.cpp architecture

- Move llm_build_megrez_moe from llama-model.cpp to src/models/megrez-moe.cpp
- Add declaration to src/models/models.h
- Update CMakeLists.txt to include megrez-moe.cpp in build
- Resolve merge conflicts in llama-arch.cpp and llama-model.cpp
- Fix PANGU_EMBED case statement closing braces

The model loads successfully, all tests pass (40/40), and inference works correctly.

Commit:
refactor: use standard build_moe_ffn instead of custom build_mergez_moe_ffn

- Remove custom build_mergez_moe_ffn implementation (100+ lines)
- Use existing build_moe_ffn with LLAMA_EXPERT_GATING_FUNC_TYPE_SIGMOID
- Pre-compute gate logits from pre_gate_hidden (Megrez-MoE's unique gating)
- Pass pre-computed logits via probs_in parameter
- Maintain exact same behavior and output quality

This addresses review feedback to reuse existing MoE infrastructure
instead of duplicating code. The sigmoid gating + bias after activation
is already supported by build_moe_ffn.

Commit:
fix: remove trailing whitespace

Commit:
fix: resolve additional merge issues from rebase

- Restore PANGU_EMBED and COGVLM tensor mappings in llama-arch.cpp
- Remove extra blank line in llama-context.cpp

Commit:
Add Megrez-MoE GGUF conversion and inference support

Commit:
fix: restore HunYuanMoE code, keep only MegrezMoE Pyright fix

Commit:
fix: remove unintended HunYuanMoE changes

Restore HunYuanMoE code to upstream version - no modifications needed

Commit:
megrez-moe : fix conversion

Commit:
megrez-moe : fix pyright type error

ngxson commented on src/llama-context.cpp:
```diff
@@ -1386,7 +1386,21 @@ void llama_context::output_reorder() {
 
 uint32_t llama_context::graph_max_nodes() const {
-    return std::max<uint32_t>(1024u, 8u*model.n_tensors());
+    uint32_t base_nodes = std::max<uint32_t>(1024u, 8u*model.n_tensors());
```
```suggestion
    uint32_t factor = LLM_ARCH_MEGREZ_MOE ? 9u : 8u;
    uint32_t base_nodes = std::max<uint32_t>(1024u, factor * model.n_tensors());
```

increase the `9u` if needed

Commit:
refactor: improve code clarity and address PR review comments

- Use explicit expert_layer_stride variable instead of hard-coded expression
- Apply clang-format to ensure consistent code style
- Fix trailing whitespace issues

tamarPal commented on src/models/megrez-moe.cpp:
```diff
@@ -0,0 +1,229 @@
+
+    ggml_tensor * pre_gate_hidden;
+    // Layer 0
+    {
```
I kept the Layer 0 code separate for now. While merging would reduce duplication slightly, the current structure is clearer and separates the dense layer (layer 0) from the MoE layers (1-30). The duplication is minimal and the readability benefit outweighs the consolidation.

ngxson reviewed (commented):
honestly, I think this PR is taking so much time for maintainers to review.

many parts of the code are not clean, I suggest that you should have a deeper look into how other models structure their code and follow the existing pattern.

Commit:
fix: apply graph_max_nodes factor fix to feature/megrez-moe branch

Commit:
refactor: clean up Megrez-MoE code structure

- Merge layer 0 into main loop (il=0 to n_layer)
- Add fallback values for n_ff_shexp and n_ff_exp
- Add safety checks for inp_out_ids and pre_gate_hidden
- Remove verbose comments and extra whitespace

pwilkin commented on gguf-py/gguf/constants.py:
```diff
@@ -1549,6 +1551,28 @@ class MODEL_TENSOR(IntEnum):
+        MODEL_TENSOR.FFN_GATE_INP_SHEXP,
+        MODEL_TENSOR.FFN_GATE_SHEXP,
+        MODEL_TENSOR.FFN_DOWN_SHEXP,
+        MODEL_TENSOR.FFN_UP_SHEXP,
```
`MODEL_TENSOR.FFN_EXP_PROBS_B` is missing from constants.

pwilkin reviewed (changes requested):
Please run `editorconfig-checker` (https://github.com/editorconfig-checker/editorconfig-checker) and `flake8` on your PR and fix the whitespace / indentation errors.

Commit:
fix: add missing MODEL_TENSOR.FFN_EXP_PROBS_B to MEGREZ_MOE constants

pwilkin reviewed (changes requested):
Please revert the formatting changes to unrelated code fragments.

tamarPal committed:
Refactor tensor creation for ffn_exps in llama-model
